internal/hcsoci/pnp.go

```diff
@@ -0,0 +1,58 @@
+// +build windows
+
+package hcsoci
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Microsoft/hcsshim/internal/log"
+	"github.com/Microsoft/hcsshim/internal/shimdiag"
+	"github.com/Microsoft/hcsshim/internal/uvm"
+	"github.com/pkg/errors"
+)
+
+const uvmPnpExePath = "C:\\Windows\\System32\\pnputil.exe"
+
+// createPnPInstallDriverCommand creates a pnputil command to add and install drivers
+// present in `driverUVMPath` and all subdirectories.
+func createPnPInstallDriverCommand(driverUVMPath string) []string {
```

Comment:
What happens if the driver has already been installed for a different container request?

Reply:
If one container is running with version X.X of a driver and a second container comes along and requests a different directory containing the same driver but using version X.Y (where Y>X), the second driver will overwrite the first driver in pnp since the version is higher. We have no way of preventing that. If the second container uses the same version, nothing happens, pnp will bail out on installation early. If the second container requests a lower version of the same driver, that installation will fail, since a "better" driver exists on the system already.

Reply:
I assume we aren't aware of any use case where this type of scenario will occur? If so, this should be fine to leave as-is for now, but let's make sure this restriction is documented.

Reply:
I thought I had remembered we needed to add "/wait" or something to the pnputil call to avoid race conditions in the case of 2 containers using the same driver. Is that not needed?

Reply:
I don't recall a worry about different driver versions between containers. IIRC, we just need to make sure that we install drivers before assigning devices to containers so that all devices in the UVMs object namespaces are setup. When pnputil returns from an install command, we know that the device is ready to use.